### DISCUSSION TITLE
Run using Cypress module API

### DISCUSSION THREAD

**Discussion by abhidp** [sorry-cypress/sorry-cypress#421] [Q&A] [answered]
@agoldis 
We have a lot of custom logic and configuration for which we use [Cypress Module API](https://docs.cypress.io/guides/guides/module-api) to run the tests as a node process, not a `cypress run` binary process. 

How do we use our existing module API node process to run sorry-cypress?

**Comment by agoldis** [comment, marked as answer]
@abhidp Checkout out `patch` option of `cy2` package https://www.npmjs.com/package/cy2.
It'd work for your use case

**Reply by abhidp** [reply]
Sorry @agoldis . The documentation of `cy2` is very sparse and not very informative. 
Can you please provide a detailed working example of how to convert the following cypress module api node program to cy2 ? 

```javascript
//cypress.run.js

const cypress = require('cypress')

const options = {
  headless: false,
  headed: true,
  browser: 'chrome',
  'no-exit': true,
  spec: 'cypress/integration/groupA/a.spec.js',
  config: {
    integrationFolder: `cypress/integration`,
    video: false
  },
  env: {
    FOO: 'bar',
    BAZ: 'cux'
  }
}

const cypressRun = async () => {
  const results = await cypress.run(options)
}

cypressRun().catch(console.error)
```

the above program is run using `node cypress.run.js`

please provide a corresponding example for cy2, lets say in a file called `cy2.run.js` which can be run using a node process like so `node cy2.run.js`

**Reply by agoldis** [reply]
Just invoke `patch` method from `cy2`

```js
const { patch } = require('cy2');
const cypress = require('cypress')

const options = {
  headless: false,
  headed: true,
  browser: 'chrome',
  'no-exit': true,
  spec: 'cypress/integration/groupA/a.spec.js',
  config: {
    integrationFolder: `cypress/integration`,
    video: false
  },
  env: {
    FOO: 'bar',
    BAZ: 'cux'
  }
}

const cypressRun = async () => {
  await patch('https://sorry-cypress-demo-director.herokuapp.com');
  const results = await cypress.run(options)
}

cypressRun().catch(console.error)
```

**Reply by abhidp** [reply]
thanks @agoldis , I have sent you a coffee.

**Reply by agoldis** [reply]
Thanks @abhidp that's very nice of you :)

**Reply by sdharmik** [reply]
@agoldis, I am running cypress automation using cypress module api as below. As suggested by you, I have used the patch method of sorry cypress however the result is not getting published on Sorry Cypress dashboard.
Could you please help on it?

const cypress = require('cypress')
const { patch } = require('cy2');

const options = {    
    env: {
      environment: 'qa'
    }
  }
const cypressRun = async () => {
    await patch('https://sorry-cypress-director.com');
    const results = await cypress.run(options)
  }
  
  cypressRun().catch(console.error)

**Reply by agoldis** [reply]
https://docs.sorry-cypress.dev/configuration/troubleshooting#why-isnt-sorry-cypress-dashboard-working-i-cannot-see-test-results

**Reply by sdharmik** [reply]
@agoldis Thanks for your quick reply. I have gone through the given link. 

When executed below command from CLI, result gets published on Sorry Cypress dashboard.
export CYPRESS_API_URL="https://sorry-cypress-director.com/"
node_modules/.bin/cy2 run --parallel --record --key somekey --ci-build-id build9

However we use Cypress Module API to run the tests as a node process and with the suggested patch option, when I am running automation, result is not redirected on sorry cypress dashboard. Also what will be the build id when using the patch option?

const cypressRun = async () => {
  await patch('https://sorry-cypress-demo-director.herokuapp.com');
  const results = await cypress.run(options)
}

can you please help on it?

**Reply by sdharmik** [reply]
@agoldis I am eagerly awaiting your positive response. As I mentioned earlier when executed below sorry cypress command from CLI, result is published to sorry cypress dashboard. However I want to make it happen through cypress.run.js file as I am using Cypress module api and suggested patch method is not working for me to publish the result on dashboard.
Your input is very important to us.

export CYPRESS_API_URL="https://sorry-cypress-director.com/"
node_modules/.bin/cy2 run --parallel --record --key somekey --ci-build-id build9

**Reply by agoldis** [reply]
@sdharmik

Most chances something is wrong with your setup or the way you're connecting to sorry-cypress. Try following the next steps for troubleshooting before submitting support requests.

Ensure you're [reconfiguring cypress](https://github.com/cypress-agent/cy2) to use Director's service URL. 
- Run cypress in debug mode `DEBUG=cypress:server:*` to see the details of network requests - the debug mode works both when running `cypress` and `cy2`
- Examine Director service log files to see the incoming requests
